pkg/compactor/block_upload.go

```diff
@@ -36,62 +36,91 @@ const uploadingMetaFilename = "uploading-" + block.MetaFilename
-	vars := mux.Vars(r)
-	blockID := vars["block"]
-	bULID, err := ulid.Parse(blockID)
-	if err != nil {
-		http.Error(w, "invalid block ID", http.StatusBadRequest)
+func (c *MultitenantCompactor) StartBlockUpload(w http.ResponseWriter, r *http.Request) {
+	blockID, tenantID, userBkt, ok := c.parseBlockUploadParameters(w, r)
+	if !ok {
 		return
 	}
+
 	ctx := r.Context()
-	tenantID, err := tenant.TenantID(ctx)
-	if err != nil {
-		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+	logger := log.With(util_log.WithContext(ctx, c.logger), "block", blockID)
+
+	const op = "start block upload"
+
+	if err := checkForCompleteBlock(ctx, blockID, userBkt); err != nil {
+		writeBlockUploadError(err, op, "while checking for complete block", logger, w)
 		return
 	}
-	if !c.cfgProvider.CompactorBlockUploadEnabled(tenantID) {
-		http.Error(w, "block upload is disabled", http.StatusBadRequest)
+
+	if err := c.createBlockUpload(ctx, r, logger, userBkt, tenantID, blockID); err != nil {
+		writeBlockUploadError(err, op, "", logger, w)
 		return
 	}
 
-	logger := log.With(util_log.WithContext(ctx, c.logger), "block", blockID)
+	w.WriteHeader(http.StatusOK)
+}
 
-	shouldComplete := r.URL.Query().Get("uploadComplete") == "true"
-	var op string
-	if shouldComplete {
-		op = "complete block upload"
-	} else {
-		op = "start block upload"
+// FinishBlockUpload handles request for finishing block upload.
+//
+// Finishing block upload performs block valiation, and if all checks pass, marks block as finished
+// by uploading meta.json file.
+func (c *MultitenantCompactor) FinishBlockUpload(w http.ResponseWriter, r *http.Request) {
+	blockID, _, userBkt, ok := c.parseBlockUploadParameters(w, r)
+	if !ok {
+		return
 	}
 
-	userBkt := bucket.NewUserBucketClient(tenantID, c.bucketClient, c.cfgProvider)
-	if err := checkForCompleteBlock(ctx, bULID, userBkt); err != nil {
+	ctx := r.Context()
+	logger := log.With(util_log.WithContext(ctx, c.logger), "block", blockID)
+
+	const op = "complete block upload"
+
+	if err := checkForCompleteBlock(ctx, blockID, userBkt); err != nil {
 		writeBlockUploadError(err, op, "while checking for complete block", logger, w)
 		return
 	}
 
-	if shouldComplete {
-		err = c.completeBlockUpload(ctx, r, logger, userBkt, bULID)
-	} else {
-		err = c.createBlockUpload(ctx, r, logger, userBkt, tenantID, bULID)
-	}
-	if err != nil {
+	if err := c.completeBlockUpload(ctx, r, logger, userBkt, blockID); err != nil {
 		writeBlockUploadError(err, op, "", logger, w)
 		return
 	}
 
 	w.WriteHeader(http.StatusOK)
 }
 
+// parseBlockUploadParameters parses common parameters from the request: block ID, tenant and checks if tenant has uploads enabled.
+// If it returns ok==false, then HTTP response has been sent.
+func (c *MultitenantCompactor) parseBlockUploadParameters(w http.ResponseWriter, r *http.Request) (blockULID ulid.ULID, tenantID string, userBucket objstore.InstrumentedBucket, ok bool) {
```

Comment:
The pattern of this method returning values _and_ writing to the HTTP response seems awkward to me. My preference would be for it to return `(blockULID, tenantID, userBucket, error)` and have the caller write the error as a response.

Reply:
From a design perspective, I agree with Nick. Returning error calling `http.Error()` in the 3 callers means 3 more lines of code, which looks non significant to me given the code will be more clear.

Reply:
OK, two votes against one, you win. Also removing bucket client creation.